### PR TITLE
[stdlib] Implement native normalization for String

### PR DESCRIPTION
**Pull request by Azoy**
This is still a wip, but wanted to get some benchmarks and tests ran on this.

Resolves: rdar://51635207, https://bugs.swift.org/browse/SR-9432

### CODE REVIEW

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please benchmark

File: stdlib/public/core/String.swift [hunk 1]
```diff
@@ -982,11 +982,105 @@ extension String {
 
 extension _StringGutsSlice {
   internal func _withNFCCodeUnits(_ f: (UInt8) throws -> Void) rethrows {
-    for scalar in Substring(self).nfc {
-      try scalar.withUTF8CodeUnits {
-        for byte in $0 {
+    // Fast path: If we're already NFC (or ASCII), then we don't need to do
+    // anything at all.
+    if _guts.isNFC {
```
Comment (Catfish-Man) [inline stdlib/public/core/String.swift:987]: Something to consider for patterns like this is splitting the function in two, so that the fast path can be inlined into callers. I don't know what the state of partial inlining in swiftc is but I haven't found it to be something I can rely on.

**Comment by karwa** [comment]
Awesome! I have a couple of comments though:

1. It would be _really_ valuable if we could make this available to non-String collections of unicode text.
2. Have you considered perhaps not decoding the UTF-8 code-units in to full scalars, and leaving them in their encoded form? I've found that decoding to be surprisingly expensive -- yet if you handle/don't care about overlong encodings, there is a 1:1 mapping between an "encoded scalar" and its decoded value, and thanks to the way UTF-8 works, they sort the same way. That means they are usable in lookup tables and it is possible to match ranges of scalars by bit-masking the UTF-8-encoded version.

Finally, as excited as I am to see progress here, my excitement is dampened (_just a liiiitle bit_) by the use of C. Is it really not possible to write these parts in Swift? Even using `withUnsafeBufferPointer` to bypass bounds-checking?

**Comment by milseman** [comment]
> 1. It would be really valuable if we could make this available to non-String collections of unicode text.

Yes, a huge benefit of native normalization is exposing better APIs in the future. Right now the critical part is nailing down the implementation and dropping ICU, but with that we can add normalization (and grapheme breaking) APIs that take collections of scalars or code units.

Another important API for performance will be some kind of `makeNormal()` call, normalizing initializers, etc., that will put the string into the stdlib's preferred comparison form (which happens to be NFC).

> 2. Have you considered perhaps not decoding the UTF-8 code-units in to full scalars, and leaving them in their encoded form?

Fast-paths are vital for normalization performance and give us the biggest bang for our complexity budget. The vast majority of the time you don't have to actually do any work, and fast paths are about finding that quickly. We should try to detect the biggest ones directly in UTF-8. IIRC Alejandro has a fast path for a leading byte < `0xCC`, which catches all scalars with values less than 0x300 (the first combining scalar). This is by far and away the most important fast path (other than the internal `isNFC` perf flag I suppose), as the most frequent strings for comparison are not linguistic text but rather quasi-human/machine readable strings. This also covers the vast majority of Western European and some African and South-East Asian language text too, which is a nice benefit.

The next fast-path would be leading byte checks that span the largest and most important ranges of NFC scalars, such as the Han ideograph code pages. (Finally, Arabic and Cyrillic checks would help round out normally-precomposed languages).

We do need to weigh against complexity. Beyond these fast-paths, I think it makes sense for the general purpose fall-back path to be based on scalars. We need that for lazy UTF-16 Cocoa strings and future API anyways. Any extra effort should probably be channeled into the fast-paths and APIs at that point.

**Comment by Azoy** [comment]
> Finally, as excited as I am to see progress here, my excitement is dampened (_just a liiiitle bit_) by the use of C. Is it really not possible to write these parts in Swift? Even using `withUnsafeBufferPointer` to bypass bounds-checking?

The use of C for storing Unicode data and accessing it is simply due to the fact that these data structures need to live within the binary. If I were to store these in Swift, either as a tuple or an array, each element would need to be initialized on program launch. The Swift compiler can't emit immutable globals as data into the binary yet, so writing at least the data structures in C is the only choice at the moment. As for accessing said data, it's simply that importing these C fixed sized arrays are gross to work with in Swift and is easier to access the data from the source language its defined in.

**Comment by karwa** [comment]
> The use of C for storing Unicode data and accessing it is simply due to the fact that these data structures need to live within the binary. If I were to store these in Swift, either as a tuple or an array, each element would need to be initialized on program launch. The Swift compiler can't emit immutable globals as data into the binary yet, so writing at least the data structures in C is the only choice at the moment. 

Ahhh that's right, I forgot about that; looking at things in godbolt, it does indeed allocate and populate immutable globals at launch :(. I couldn't find a bug report on bugs.swift.org about it; do you know if it's being tracked? Otherwise I'd be happy to file one.

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please benchmark

File: stdlib/public/core/NFC.swift [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    }
+    
+    switch (x.value, y.value) {
+    // Check for Hangul (L, V) -> LV compositions.
+    case (0x1100 ... 0x1112, 0x1161 ... 0x1175):
+      let lIdx = x.value &- 0x1100
+      let vIdx = y.value &- 0x1161
+      let lvIdx = lIdx &* 588 &+ vIdx &* 28
+      let s = 0xAC00 &+ lvIdx
+      return Unicode.Scalar(s)
```
Comment (karwa) [inline stdlib/public/core/NFC.swift:51]: Why do some of these use `Unicode.Scalar(_value:)` and others use the failable initializer?

File: stdlib/public/core/NFC.swift [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    }
+    
+    // If we have a leftover composee, make sure to return it.
+    return composee.take()
+  }
+}
+
+extension Unicode.NFC: Sequence {
+  internal func makeIterator() -> Iterator {
+    Iterator(iterator: base.nfd.makeIterator())
```
Comment (karwa) [inline stdlib/public/core/NFC.swift:193]: Is it perhaps worth reserving capacity in the buffer? Otherwise it will be the empty array singleton, and the first append will allocate IIRC.

File: stdlib/public/core/NFD.swift [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    
+    // This size is the utf8 length of the decomposition.
+    var size = Int(decompPtr[decompIdx])
+    
+    let utf8 = UnsafeBufferPointer(
+      start: decompPtr + decompIdx + 1,
+      count: size
+    )
+
+    while size != 0 {
```
Comment (karwa) [inline stdlib/public/core/NFD.swift:105]: If this can't be negative, I'd recommend `size > 0`

File: stdlib/public/core/NFD.swift [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    // rather than removing from the front.
+    buffer.reverse()
+    
+    return buffer.removeLast()
+  }
+}
+
+extension Unicode.NFD: Sequence {
+  internal func makeIterator() -> Iterator {
+    Iterator(iterator: base.unicodeScalars.makeIterator())
```
Comment (karwa) [inline stdlib/public/core/NFD.swift:212]: Worth reserving buffer capacity?

File: stdlib/public/core/String.swift [hunk 6]
```diff
@@ -981,115 +981,110 @@ extension String {
...
-    for i in 0..<result.amountFilled {
-      try f(outputBuffer[i])
+    var isNFCQC = true
+
+    if _guts.isFastUTF8 {
+      _guts.withFastUTF8 { utf8 in
+        var position = 0
+        var prevCCC: UInt16 = 0
+
+        while position != utf8.count {
```
Comment (karwa) [inline stdlib/public/core/String.swift:1032]: `<` is usually a better choice than `!=`

**Review by karwa — COMMENTED** [review]
Minor nits, but it looks good! Really looking forward to it!

Also, I wonder if there is a way to do `Array.removeLast()` while guaranteeing the existing capacity is kept...

File: stdlib/public/core/NFC.swift (hunk 3 above)
Comment (Azoy) [reply stdlib/public/core/NFC.swift:193]: So, I experimented with this actually, and what I found was that reserving capacity for all strings didn't make sense. The reason for this is that a lot of the common strings don't even go use the buffer at all, so reserving capacity for this buffer made some of the benchmarks slower for things like ascii, latiny strings, etc. For strings that do use the buffer, this provided little to no difference in terms of benchmarking performance. I only ever saw negligible wins with our zalgo benchmarks when I did something crazy like `reserveCapacity(100)`.

**Comment by milseman** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please clean test Linux

File: stdlib/public/core/NFD.swift [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    while let scalar = iterator.next() {
+      let normData = _swift_stdlib_getNormData(scalar.value)
+      let ccc = normData >> 3
+      var isNFDQC = normData & 1 == 0
+      
+      // Because we don't store precomposed hangul in our NFD_QC data, these
+      // will return true for NFD_QC when in fact they are not.
+      if (0xAC00 ... 0xD7A3).contains(scalar.value) {
+        isNFDQC = false
+      }
```
Comment (milseman) [inline stdlib/public/core/NFD.swift:168]: It would be really nice to refactor this into the computed property on `_NormData`.

File: stdlib/public/core/NFD.swift [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+      
+      // If we've reached a starter, stop.
+      if ccc == 0, !buffer.isEmpty {
+        starter = (scalar, normData)
+        break
+      }
+      
+      // If our scalar IS NFD quick check, then it's as simple as appending to
```
Comment (milseman) [inline stdlib/public/core/NFD.swift:176]: This is copy-pasta. Is there a way to refactor it?

File: stdlib/public/core/NFD.swift [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+    }
+    
+    // Sort the entire buffer based on the canonical combining class.
+    buffer._insertionSort(within: buffer.indices) {
+      ($0.normData >> 3) < ($1.normData >> 3)
+    }
+    
+    // Reverse the buffer after sorting to make removals faster from the back
+    // rather than removing from the front.
+    buffer.reverse()
```
Comment (milseman) [inline stdlib/public/core/NFD.swift:204]: Again, a read index is probably clearer. You can even make a `_NormDataBuffer` struct that holds the array and the read index and can just encapsulate the fundamental operations.

File: stdlib/public/core/NFD.swift [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+extension Unicode.NFD: Sequence {
+  internal func makeIterator() -> Iterator {
+    Iterator(iterator: base.unicodeScalars.makeIterator())
+  }
+}
+
+extension StringProtocol {
+  internal var nfd: Unicode.NFD<Self> {
```
Comment (milseman) [inline stdlib/public/core/NFD.swift:217]: Probably underscore

File: stdlib/public/core/NFD.swift [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+extension Unicode {
+  internal struct NFD<S: StringProtocol> {
```
Comment (milseman) [inline stdlib/public/core/NFD.swift:16]: Underscore

File: stdlib/public/core/Optional.swift [hunk 12]
```diff
@@ -262,6 +262,32 @@ public enum Optional<Wrapped>: ExpressibleByNilLiteral {
...
+  ///       print(numberOfShoes)
+  ///       // Prints "34"
+  ///     }
+  ///
+  ///     print(numberOfShoes)
+  ///     // Prints "nil"
+  ///
+  /// - Returns: The wrapped value being stored in this instance. If this
+  ///   instance is `nil`, returns `nil`.
+  internal mutating func take() -> Wrapped? {
```
Comment (milseman) [inline stdlib/public/core/Optional.swift:282]: underscore it, especially if you're looking at making API in this area.

File: stdlib/public/core/String.swift [hunk 13]
```diff
@@ -981,115 +981,119 @@ extension String {
...
-    } else {
-      return try _foreignWithNormalizedCodeUnitsImpl(
-        outputBuffer: _castOutputBuffer(&output),
-        icuInputBuffer: _castOutputBuffer(&icuInput),
-        icuOutputBuffer: _castOutputBuffer(&icuOutput),
-        f
-      )
+  internal func _forEachUTF8(_ f: (UInt8) throws -> Void) rethrows {
+    for byte in String(_guts).utf8 {
+      try f(byte)
```
Comment (milseman) [inline stdlib/public/core/String.swift:986]: Is this `String(_guts).utf8.forEach(f)`?

File: stdlib/public/core/String.swift [hunk 14]
```diff
@@ -981,115 +981,119 @@ extension String {
...
+  ) -> Bool {
+    let normData: UInt16
 
-    var index = range.lowerBound
-    let cachedEndIndex = range.upperBound
+    if scalar.value < 0x300 {
+      normData = 0
+    } else {
+      normData = _swift_stdlib_getNormData(scalar.value)
+    }
```
Comment (milseman) [inline stdlib/public/core/String.swift:1000]: Probably better as an initializer or static method on `_NormData`.

File: stdlib/public/core/String.swift [hunk 15]
```diff
@@ -981,115 +981,119 @@ extension String {
...
-  let cachedEndIndex = String.Index(_encodedOffset: sourceBuffer.count)
-
-  var hasBufferOwnership = false
-
-  defer {
-    if hasBufferOwnership {
-      outputBuffer.deallocate()
-      icuInputBuffer.deallocate()
-      icuOutputBuffer.deallocate()
+  internal func _withNFCCodeUnits(_ f: (UInt8) throws -> Void) rethrows {
```
Comment (milseman) [inline stdlib/public/core/String.swift:1017]: Maybe "forEachNFCCodeUnit"

Comment (milseman) [reply stdlib/public/core/String.swift:1017]: We're going to want to be looking to hand-outline any code that doesn't depend on `f` that we can. Makes it easier to inline `f` into this or this into the caller (which is unfortunately sometimes necessary to inline `f`).

File: stdlib/public/core/String.swift [hunk 16]
```diff
@@ -981,115 +981,119 @@ extension String {
...
+    // Fast path: If we're already NFC (or ASCII), then we don't need to do
+    // anything at all.
+    if _guts.isNFC {
```
Comment (milseman) [inline stdlib/public/core/String.swift:1020]:
```suggestion
    if _fastPath(_guts.isNFC) {
```

File: stdlib/public/core/String.swift [hunk 17]
```diff
@@ -981,115 +981,119 @@ extension String {
...
+          let (scalar, len) = _decodeScalar(utf8, startingAt: position)
+
+          if !_isScalarNFCQC(scalar, &prevCCC) {
+            isNFCQC = false
+            return
+          }
+
+          position &+= len
+        }
+      }
```
Comment (milseman) [inline stdlib/public/core/String.swift:1059]: Try refactoring/outlining this whole thing. Ideally, we'd have a function call, then just the loop invoking `f`.

File: stdlib/public/core/String.swift [hunk 18]
```diff
@@ -981,115 +981,119 @@ extension String {
...
-    if result.allocatedBuffers {
-      _internalInvariant(!hasBufferOwnership)
-      hasBufferOwnership = true
+
+    for scalar in String(_guts).nfc {
+      try scalar.withUTF8CodeUnits {
+        for byte in $0 {
+          try f(byte)
+        }
+      }
```
Comment (milseman) [inline stdlib/public/core/String.swift:1096]: If we want to make a public `String.nfc` var, then all this code should probably be in its iterator's `next()` method.

File: stdlib/public/core/StringComparison.swift [hunk 19]
```diff
@@ -247,7 +247,10 @@ private func _findBoundary(
     if idx == 0 { return 0 }
 
     let scalar = _decodeScalar(utf8, startingAt: idx).0
-    if scalar._hasNormalizationBoundaryBefore { return idx }
+
+    if scalar._isNFCStarter {
+      return idx
+    }
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift:253]: Why this change?

File: stdlib/public/stubs/UnicodeData.cpp [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "../SwiftShims/UnicodeData.h"
+
+static inline __swift_uint32_t scramble(__swift_uint32_t scalar) {
```
Comment (milseman) [inline stdlib/public/stubs/UnicodeData.cpp:15]: comment please

File: stdlib/public/stubs/UnicodeData.cpp [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+  return hash % level;
+}
+
+SWIFT_RUNTIME_STDLIB_INTERNAL
+__swift_intptr_t _swift_stdlib_getMphIdx(__swift_uint32_t scalar,
+                                         __swift_intptr_t levels,
+                                         const __swift_uint64_t * const *keys,
+                                         const __swift_uint16_t * const *ranks,
+                                         const __swift_uint16_t * const sizes) {
+  __swift_intptr_t resultIdx = 0;
```
Comment (milseman) [inline stdlib/public/stubs/UnicodeData.cpp:51]: Definitely add comments, and think about citations if relevant.

**Review by milseman — COMMENTED** [review]
Quick code review. Mostly I'm interested in small refactoring to make the implementation easier to read and understand.

File: stdlib/public/core/NFC.swift [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+          buffer.append(current)
+          continue
+        }
+        
+        // Starters are the "start" of a new normalization segment. Set it as
+        // the new composee and return our current composee. This will trigger
+        // any other scalars in the buffer to be emitted before we handle
+        // composing this new composee.
+        composee = current.scalar
+        return l
```
[the remark this answers is not in the document]
Comment (Azoy) [reply stdlib/public/core/NFC.swift:167]: Heh, I was using Unicode's names that they use in the normalization conformance section, for example:

> If there is such an L, and C is not blocked from L, and there exists a Primary Composite P which is canonically equivalent to the sequence <L, C>, then replace L by P in the sequence and delete C from the sequence.

but yes, better names can be found.

File: stdlib/public/core/NFD.swift [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+      
+      let tIdx = sIdx % 28
+      if tIdx != 0 {
+        let t = Unicode.Scalar(_value: 0x11A7 &+ tIdx)
+        // Hangul tail consonants, T, always have normData of 4.
+        // CCC = 0, NFC_QC = Maybe, NFD_QC = Yes
+        buffer.append((scalar: t, normData: 4))
+      }
+      
+      return
```
[the remark this answers is not in the document]
Comment (Azoy) [reply stdlib/public/core/NFD.swift:68]: Does it make sense to also annotate `decomposeSlow` as `@inline(never)` as well to prevent inlining that into `decompose`?

File: stdlib/public/SwiftShims/UnicodeShims.h [hunk 24]
```diff
@@ -489,27 +488,6 @@ SWIFT_RUNTIME_STDLIB_API
...
-__swift_stdlib_unorm2_normalize(const __swift_stdlib_UNormalizer2 *,
-                                const __swift_stdlib_UChar *, __swift_int32_t,
-                                __swift_stdlib_UChar *, __swift_int32_t,
-                                __swift_stdlib_UErrorCode *);
-
-SWIFT_RUNTIME_STDLIB_API
-__swift_int32_t __swift_stdlib_unorm2_spanQuickCheckYes(
-    const __swift_stdlib_UNormalizer2 *, const __swift_stdlib_UChar *,
-    __swift_int32_t, __swift_stdlib_UErrorCode *);
-
```
[the remark this answers is not in the document]
Comment (Azoy) [reply stdlib/public/SwiftShims/UnicodeShims.h:512]: I'm not sure what you mean by this, do these definitions have declarations? If so, yes they do, and git doesn't do a great job of showing that the current file where these are in are `UnicodeNormalization.cpp` but I wanted to use this name for the native stuff, so I moved the ICU declarations in `UnicodeShims.cpp` and removed these.

[the remark this answers is not in the document]
Comment (milseman) [reply stdlib/public/SwiftShims/UnicodeShims.h:512]: I was referring to the contents of `stdlib/public/SwiftShims/UnicodeShims.h`. If we're removing stuff, we probably want to remove the `.h` file entry too so that this can be a compile-time error instead of a link-time error.

File: stdlib/public/core/NFC.swift [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+    }
+    
+    switch (x.value, y.value) {
+    // Check for Hangul (L, V) -> LV compositions.
+    case (0x1100 ... 0x1112, 0x1161 ... 0x1175):
+      let lIdx = x.value &- 0x1100
+      let vIdx = y.value &- 0x1161
+      let lvIdx = lIdx &* 588 &+ vIdx &* 28
+      let s = 0xAC00 &+ lvIdx
+      return Unicode.Scalar(_value: s)
```
[the remark this answers is not in the document]
Comment (Azoy) [reply stdlib/public/core/NFC.swift:51]: I can't remember if we already have tests that test this path, but I'm updating the normalization test to use Unicode's updated normalization tests which do exercise this path.

File: stdlib/public/core/StringComparison.swift (hunk 19 above)
Comment (Azoy) [reply stdlib/public/core/StringComparison.swift:253]: There was another change that I did somewhere that prompt this, but it seems tests pass either way with this change. Is there any reason tests wouldn't exercise something here that it still makes sense to revert?

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Review by milseman — COMMENTED** [review]
This is coming along very well! Just some minor feedback

File: stdlib/public/core/NFC.swift [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+      //     "z\u{0335}\u{0327}\u{0324}\u{0301}"
+      //
+      // In this example, there are several combining marks following a 'z', but
+      // none of them actually compose with the composee 'z'. However, the last
+      // scalar U+0301 does actually compose. So this check makes sure that the
+      // last scalar doesn't have any scalar in between it and the composee that
+      // would otherwise "block" it from composing.
+      guard lastBufferedNormData.ccc < current.normData.ccc else {
+        // We had a scalar block it. That means our current scalar has the same
+        // class as the previous one (preserve ordering). Stick it at the end of
```
[the remark this answers is not in the document]
Comment (Azoy) [reply stdlib/public/core/NFC.swift:177]: Well, actually it means the scalar is `>=`, so the comment here is wrong. I'll update it.

File: stdlib/public/core/String.swift [hunk 27]
```diff
@@ -981,115 +981,106 @@ extension String {
...
 
-    var index = range.lowerBound
-    let cachedEndIndex = range.upperBound
+    var isNFCQC = true
+    var prevCCC: UInt8 = 0
 
-    var hasBufferOwnership = false
+    if _guts.isFastUTF8 {
+      _fastNFCCheck(&isNFCQC, &prevCCC)
+
```
[the remark this answers is not in the document]
Comment (Azoy) [reply stdlib/public/core/String.swift:1015]: We do a single pass over the String to determine NFC, and bail out as soon as we find it isn't. Once we determine whether the String is/isn't NFC we emit every scalar from the respectful iterator.

[the remark this answers is not in the document]
Comment (milseman) [reply stdlib/public/core/String.swift:1015]: Right, but everything up to the starter prior to non-NFC content is NFC, and we could be feeding `f` that directly. I.e. we have the index of the last-seen starter, and when we see a new starter, we feed the bytes in-between to `f`. That way it's a single pass over an NFC string, and even NFC prefixes are emitted early.

Is this iterator used for a whole-String-consuming operation like hash, or a bail-early operation like comparison?

[the remark this answers is not in the document]
Comment (Azoy) [reply stdlib/public/core/String.swift:1015]: Ah yeah, I remember trying to do something like this early on, but got rid of it for some reason I can't remember. I can take a look at attempting this again.

This iterator is only used for whole consuming operations like hash (I believe hash is actually the only thing that uses it.)

[the remark this answers is not in the document]
Comment (milseman) [reply stdlib/public/core/String.swift:1015]: Ah, then it's less critical (but still good for large strings to not blow out the cache). I'd say keep this approach for now, assuming you're able to test the code path as well with the conformance suite.

**Comment by Azoy** [comment]
@swift-ci please clean test Linux

**Comment by Azoy** [comment]
@swift-ci please clean test Windows

File: stdlib/public/core/NFC.swift [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+  internal mutating func next() -> Unicode.Scalar? {
+    // Empty out our buffer before attempting to compose anything with our new
+    // composee.
+    if let nextBuffered = buffer.next() {
+      return nextBuffered.scalar
+    }
+
+    while let current = iterator.next() {
+      guard let currentComposee = composee else {
+        // If we don't have a composee at this point, we're most likely looking
```
[the remark this answers is not in the document]
Comment (Azoy) [reply stdlib/public/core/NFC.swift:110]: When the start of the string begins with say 50 non starters, we just emit those and do no normalization work, but the 51st scalar is a starter so technically we're not doing any work until that scalar.

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please test

File: stdlib/public/core/NFC.swift [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+
+      // If we have any scalars in the buffer, it means those scalars couldn't
+      // compose with our composee to form a new scalar. However, scalars
+      // following them may still compose with our composee, so take the last
+      // scalar in the buffer and get its normalization data so that we can
+      // perform the check underneath this one about whether this current scalar
+      // is "blocked". We get the last scalar because the scalars we receive are
+      // already NFD, so the last scalar in the buffer will have the highest
+      // CCC value in this normalization segment.
+      guard let lastBufferedNormData = buffer.last?.normData else {
```
Comment (milseman) [inline stdlib/public/core/NFC.swift:130]: It took me a few readings to figure it out, but we're really checking the prior scalar's CCC, right? So maybe:

```suggestion
guard let priorCCC = buffer.last?.normData.ccc else {
```

And update the comment appropriately.

Comment (Azoy) [reply stdlib/public/core/NFC.swift:130]: It's not necessarily the prior scalar's CCC, no. I can say something like `lastBufferedCCC` however and add the `.ccc` here.

Comment (milseman) [reply stdlib/public/core/NFC.swift:130]: Isn't the prior CCC what's required by Unicode? When is it a different scalar?

"last" sounds like it's the last scalar in the current normalization segment, and thus would probably appear after `current`.

**Comment by MaxDesiatov** [comment]
Does the fact that this was merged mean that ICU dependency is going to be removed at some point?

**Comment by milseman** [comment]
Hopefully soon
